solr/contrib/ltr/src/java/org/apache/solr/ltr/LTRScoringQuery.java

```diff
@@ -223,6 +223,7 @@ public ModelWeight createWeight(IndexSearcher searcher, ScoreMode scoreMode, flo
     else{
       createWeightsParallel(searcher, scoreMode.needsScores(), featureWeights, features);
     }
+    // TODO: potential prefetchFields computation site?
```

Comment:
Hi @tomglk. I had another idea! This pull request is fun :-)

The idea unexpectedly appeared triggered by the earlier insights:

> ... re-computation of the joint prefetchFields ... could be for an established actively used feature store. concurrent ...

> ... Might `ltr.PrefetchingFieldValueFeature.exclusions=aaa,bbb,ccc,ddd,eee,fff` have other uses too? ...

> ... `disablePrefetchingFieldValueFeature` ... `ltr.PrefetchingFieldValueFeature.exclusions` ...

ManagedFeatureStore computes the joint `prefetchFields` approach:
* (+) the feature store content provides visibility into the feature store's overall prefetch fields set
* (+/-) concurrency needs to be considered when updating the prefetch fields of actively used features
* (-) the prefetch fields used may be 'too wide' i.e. all features in the store vs. all features used by the model

LTRScoringQuery computes the joint `prefetchFields` approach -- the new idea:
* (-) or (+/-) less or different visibility into the overall prefetch fields set (no visibility via the feature store but still visibility when an exception is thrown)
* (+) no concurrency concerns since the joint prefetch fields set is propagated to not-yet-used PrefetchingFieldValueFeatureWeight objects
* (+) the used prefetch fields used is 'precise' i.e. all features in the store vs. all features used by the model

pseudo code:

```
for (FeatureWeight fw : featureWeights) {
  if (fw instanceof PrefetchingFieldValueFeatureWeight) {
    PrefetchingFieldValueFeatureWeight pfvfw = (PrefetchingFieldValueFeatureWeight)fw;
    prefetchingFeatureWeights.add(pfvfw);
    prefetchFields.add(pfvfw.getField());
  }
}
for (PrefetchingFieldValueFeatureWeight pfvfw : prefetchingFeatureWeights) {
  prefetchFields.addPrefetchFields(prefetchFields);
}
```

What do you think?

Reply:
Hi @cpoerschke ,
I am glad that this PR sparks your creativity! :)
This is why I find it so valuable to have sometimes long conversations. Hearing different opinions or concerns or having to explain some thoughts makes us reevaluate the current point of view which can be really inspiring!
I am also having a lot of fun discovering more and more of the internal workings of Solr & Lucene.

Thanks for leaving the `TODO` in the code!
>  less or different visibility into the overall prefetch fields set

I think we could still provide the visibility by hooking into `doGet()` of the `ManagedFeatureStore` and collecting all the fields. So that shouldn't even be a problem.

I can clearly follow your pros and cons. My only concern is the re-computation per query. If we do the collection in the ManagedFeatureStore, we have to iterate and set the fields approx. once per shard*. If we do it after creating the weights, we have to do it once per query.
(*That's what a very quick "debugging" of the procedures while executing `TestCacheInteractionOfPrefetchingFieldValueFeature.testSimpleQuery()` with different shard numbers told me.)

I would like to think about the cost of the re-computation vs. the cost of handling the concurrency.

Reply:
> ... My only concern is the re-computation per query. ...

Excellent point! And it got me thinking about additional alternative approaches and so here's another one for the list, one _without_ the per-query re-computation but _with_ the per-query precision.

LTRScoringModel (also) computes joint `prefetchFields` approach:

* (+) the feature store content provides visibility into the feature store's overall prefetch fields set
* (+) the used prefetch fields used is 'precise' i.e. all features in the store vs. all features used by the model
* (+) thrown exceptions can provide visibility into the 'precise' prefetch fields used
* (+/-) concurrency needs to be considered when updating the prefetch fields of actively used features

partial code illustration: https://github.com/cpoerschke/solr/commit/de4e32fa21b9daba34b6b297e770e1c2c29362b0

Reply:
Hi @cpoerschke ,
from the approaches that we discussed until now, I like this one best.
I committed your implementation in 7ac6e35.
(Also, I did a few small changes to fix the tests and some renaming because the terms `frozen` and `freeze` made it a bit hard to understand.)

Reply:
> ... Also, I did a few small changes to fix the tests ...

Thanks! I'm not sure if the `setField` and `setIndex` calls in the `clone()` method were a key part to fix the tests but seeing them made me wonder if we need to give a clone method to the super class or declare the class(es) `Cloneable` since both field and index are not specific to the PFVF class.

> ... some renaming because the terms `frozen` and `freeze` made it a bit hard to understand.)

Yes, those terms were a bit generic. Upon further thought it seems that Java Collection's _unmodifiable_ concept is a good match here -- https://github.com/apache/solr/pull/166/commits/69c6611c467b8129bd31ea7527f091f734d1e3b2 tries that out, what do you think?

Reply:
> if the setField and setIndex calls in the clone() method were a key part to fix the tests

They were essential to make the tests work. I never heard about `Clonable`, but that sounds like we could use it here. :)

> 69c6611 tries that out, what do you think?

I can see where you are coming from, but to be honest I am not really a fan of this change.
Yes, it makes the code more comprehensive. But my problem is, that now we just throw a generic `UnsupportedOperationException` instead of providing more information about the cause like we did with the custom message.

Reply:
> ... I never heard about `Clonable`, but that sounds like we could use it here. :)

I've seen it but never used it myself as yet, so learnt something new here too :)

> Yes, it makes the code more comprehensive. But my problem is, that now we just throw a generic `UnsupportedOperationException` instead of providing more information about the cause like we did with the custom message.

Yes, it's a bit of a trade-off i.e. more concise code in the general common code path case at the expense of only a generic exception in the rare/impossible code path when an exception is thrown. From my current understanding of the code the

```
throw new UnsupportedOperationException("Feature " + name + " is in use by a model. Its prefetchingFields may not be changed!");
```

can't ever happen, right? And so if it did happen then some very strange bug or problem presumably would have arisen and whilst the name of the feature might be useful to figure out the details perhaps the "is in use by a model" claim made by the exception would not be accurate in that scenario i.e. when the impossible happens very few assumptions about the state of the systems can then be made.

Removal of the `keepFeaturesFinalAfterModelCreation` boolean also means that we don't need to worry about boolean vs. AtomicBoolean or other thread-safety for it and it being or not being included when the object is cloned.

Reply:
> can't ever happen, right? And so if it did happen then some very strange bug or problem presumably would have arisen

I think so too. That also made me struggle yesterday as I thought about testing.
Your arguments make sense to me. The additional complexity that e.g. the AtomicBoolean would provide also speaks against the use of the `keepFeaturesFinalAfterModelCreation`.

Maybe a comment could provide a bit more clarity for people who see the `prefetchFields = Collections.unmodifiableSortedSet(prefetchFields);` and a few lines below `prefetchFields.addAll(fields);`?


Regarding `Cloneable`:
I just experimented a bit with the `Cloneable` interface. It seems to only do a shallow copy which could be problematic with the prefetchFields. I used this code:
```java
public PrefetchingFieldValueFeature clone() throws CloneNotSupportedException {
    return (PrefetchingFieldValueFeature) super.clone();
}
```
and made the PrefetchingFieldValueFeature implement Cloneable